DEVELOPERS.md

```diff
@@ -0,0 +1,134 @@
+![Circus Train.](circus-train.png "Moving Hive data between sites.")
+
+# Developer's Guide
+
+## Overview
+
+This document is a collection of notes on Circus Train which have been put together to outline what some of the main classes do and how they link together. The project is pretty large and if you haven't worked on it for a while its easy to get lost! 
+These notes are meant as a helpful developers guide into Circus Train's code and how it works, but they are not completely exhaustive of all the inner workings of the project. Do feel free to add more information or detail. 
+
+## README.md
+
+First and foremost, its worth having a read through the [README.md](https://github.com/HotelsDotCom/circus-train) file. It is a pretty extensive guide containing a lot of info on the project, including how to run it and all the different configurations which can be used. 
+
+## Classes
+**Locomotive**
+
+* This is where it all begins.
+* A new `Replication` object is created using the `ReplicationFactory` and *replicate* is called on it.
+
+**ReplicationFactory**
+
+* Returns a `Replication` object. The type depends on whether the source table is partitioned or not, and the replication mode specified in the configuration file.
+
+**Replication**
+
+* Either partitioned or unpartitioned.
+* There are 4 replication modes:
+   * `FULL` ← default
+   * `FULL_OVERWRITE`
+   * `METADATA_MIRROR`
+   * `METADATA_UPDATE`
+* Uses a copier based on where the data is coming from and going to:
+   * HDFS or S3 → HDFS, uses `DistCpCopier`
+   * HDFS → S3, uses `S3MapreduceCpCopier`
+   * S3 → S3,  uses `S3S3Copier`
+      * Note: If you are replicating S3 → S3 cross account, *and* you want to assume a role in the target account (see `copier-options.assume-role` in `README.md`), then you must use `S3MapreduceCpCopier`.
+* The data is copied over first (if the mode is `FULL` or `FULL_OVERWRITE`).
+* Then the metadata of the table is updated.
+
+## Types of replication
+There are four types of replication which Circus Train can handle:
+
+* `FULL` ← default
+* `FULL_OVERWRITE`
+* `METADATA_MIRROR`
+* `METADATA_UPDATE`
+
+
+### Full Replication
+**Partitioned**
+
+If the source table has partitions then these and the corresponding data will be copied over to the replica table. After this, the metadata of the table will be updated. 
+
+Otherwise, if the source table has no partitions only the metadata of the table will be updated. 
```

Comment:
```suggestion
Otherwise, if the source table is partitioned but has no partitions containing data only the metadata of the table will be updated. 
```